File: src/main/java/ro/nicolaemariusghergu/queryit/service/impl/ProductServiceImpl.java
package ro.nicolaemariusghergu.queryit.service.impl;

import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import ro.nicolaemariusghergu.queryit.dto.ProductDto;
import ro.nicolaemariusghergu.queryit.mapper.ProductMapper;
import ro.nicolaemariusghergu.queryit.proxy.ProductProxy;
import ro.nicolaemariusghergu.queryit.repository.ProductRepository;
import ro.nicolaemariusghergu.queryit.service.ProductService;

import java.util.List;
import java.util.NoSuchElementException;

@Service
public record ProductServiceImpl(ProductRepository productRepository,
                                 ProductProxy productProxy)
        implements ProductService {

    @Override
    public ResponseEntity<ProductDto> findProductById(Long id) {
        return ResponseEntity.ok(productRepository.findById(id).stream()
                        .map(ProductMapper.INSTANCE::productToProductDto)
                        .findFirst()
                        .orElseThrow(
                                () -> new NoSuchElementException("Product does not exist!"))
                );
    }

    @Override
    public ResponseEntity<List<ProductDto>> getProducts() {
        return ResponseEntity.ok(productRepository.findAll()
                .stream()
                .map(ProductMapper.INSTANCE::productToProductDto)
                .toList());
    }

    @Override
    public ResponseEntity<Long> addProduct(ProductDto productDto) {
        productRepository.save(ProductMapper.INSTANCE.productDtoToProduct(productDto));
        return ResponseEntity.ok(productDto.getId());
    }

    @Override
    public ProductDto getProductByName(String name) {
        return ProductMapper.INSTANCE.productToProductDto(
                productRepository.findByName(name)
                        .orElseThrow(() ->
                                new NoSuchElementException("Product cannot be found"))
        );
    }

    @Override
    public ResponseEntity<Long> deleteProductById(Long id) {
        productRepository.deleteById(id);
        return ResponseEntity.ok(id);
    }

    @Override
    public ResponseEntity<ProductDto> updateProduct(ProductDto productDto) {
        ProductDto modifiedProduct = productRepository.findById(productDto.getId()).stream()
                .map(ProductMapper.INSTANCE::productToProductDto)
                .map(updatedProduct -> {
                    updatedProduct.setQuantity(productDto.getQuantity());
                    return updatedProduct;
                })
                .findAny()
                .orElseThrow(() ->
                        new NoSuchElementException("Product has not been found")
                );
        productRepository.save(
                ProductMapper.INSTANCE.productDtoToProduct(modifiedProduct));
        return ResponseEntity.ok(modifiedProduct);
    }

    @Override
    public ResponseEntity<List<ProductDto>> getProxyProducts(Long categoryId) {
        return ResponseEntity.ok(productProxy.getProducts(categoryId).stream()
                .map(ProductMapper.INSTANCE::productToProductDto)
                .toList());
    }

    @Override
    public ResponseEntity<List<ProductDto>> getProductsByCategoryId(Long categoryId) {
        return ResponseEntity.ok(productRepository.findAllByCategoryId(categoryId)
                .stream()
                .map(ProductMapper.INSTANCE::productToProductDto)
                .toList());
    }

}

File: src/main/kotlin/ro/nicolaemariusghergu/queryit/service/impl/ProductServiceImpl.kt
package ro.nicolaemariusghergu.queryit.service.impl

import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import ro.nicolaemariusghergu.queryit.dto.ProductDto
import ro.nicolaemariusghergu.queryit.gateway.ProductProxy
import ro.nicolaemariusghergu.queryit.mapper.ProductMapper
import ro.nicolaemariusghergu.queryit.model.Product
import ro.nicolaemariusghergu.queryit.repository.ProductRepository
import ro.nicolaemariusghergu.queryit.service.ProductService

@Service
class ProductServiceImpl(private val productRepository: ProductRepository,
private val productProxy: ProductProxy) : ProductService {
    override fun findProductById(id: Long): ResponseEntity<ProductDto> {
        return ResponseEntity.ok(productRepository.findById(id).stream()
                .map { product: Product -> ProductMapper.INSTANCE.productToProductDto(product) }
                .findFirst()
                .orElseThrow { NoSuchElementException("Product does not exist!") }
        )
    }

    override fun getProducts(): ResponseEntity<List<ProductDto>> {
        return ResponseEntity.ok(productRepository.findAll()
                .stream()
                .map {product: Product -> ProductMapper.INSTANCE.productToProductDto(product) }
                .toList())
    }

    override fun addProduct(productDto: ProductDto): ResponseEntity<Long> {
        productRepository.save(ProductMapper.INSTANCE.productDtoToProduct(productDto))
        return ResponseEntity.ok(productDto.id)
    }

    override fun getProductByName(name: String): ProductDto {
        return ProductMapper.INSTANCE.productToProductDto(
                productRepository.findByName(name)
                        .orElseThrow { NoSuchElementException("Product cannot be found") }
        )
    }

    override fun deleteProductById(id: Long): ResponseEntity<Long> {
        productRepository.deleteById(id)
        return ResponseEntity.ok(id)
    }

    override fun updateProduct(productDto: ProductDto): ResponseEntity<ProductDto> {
        val modifiedProduct: Product = productRepository.findById(productDto.id).stream()
                .map { product: Product ->
                    product.quantity = productDto.quantity
                    product
                }
                .findAny()
                .orElseThrow { NoSuchElementException("Product has not been found") }
        productRepository.save(modifiedProduct)
        return ResponseEntity.ok(ProductMapper.INSTANCE.productToProductDto(modifiedProduct))
    }

    override fun getProxyProducts(categoryId: Long): ResponseEntity<Set<ProductDto>> {
        return productProxy.getProductsFromMegaImageByCategory(categoryId)
    }

    override fun getProductsByCategoryId(categoryId: Long): ResponseEntity<List<ProductDto>> {
        return ResponseEntity.ok(productRepository.findAllByCategoryId(categoryId)
                .stream()
                .map(ProductMapper.INSTANCE::productToProductDto)
                .toList())
    }
}